beluga/include/beluga/algorithm/cluster_based_estimation.hpp

```diff
@@ -0,0 +1,422 @@
+      Sophus::SE2d{Sophus::SO2d{0.0}, Sophus::Vector2d{0.0, -parameters_.spatial_hash_resolution}},
+      Sophus::SE2d{Sophus::SO2d{+parameters_.angular_hash_resolution}, Sophus::Vector2d{0.0, 0.0}},
+      Sophus::SE2d{Sophus::SO2d{-parameters_.angular_hash_resolution}, Sophus::Vector2d{0.0, 0.0}},
+  };
+};
+
+/// For each cluster, estimate the mean and covariance of the states that belong to it.
+/**
+ * \tparam States Range type of the states.
+ * \tparam Weights Range type of the weights.
+ * \tparam Hashes Range type of the hashes.
+ * \param states Range containing the states of the particles.
+ * \param weights Range containing the weights of the particles.
+ * \param clusters Cluster ids of the particles.
+ * \return A vector of elements, containing the weight, mean and covariance of each cluster, in no particular order.
+ */
+template <class States, class Weights, class Clusters>
+[[nodiscard]] auto estimate_clusters(States&& states, Weights&& weights, Clusters&& clusters) {
+  using State = typename ranges::range_value_t<States>;
+  using Weight = typename ranges::range_value_t<Weights>;
+  using Cluster = typename ranges::range_value_t<Clusters>;
+
+  using EstimateState = std::decay_t<decltype(std::get<0>(beluga::estimate(states, weights)))>;
+  using EstimateCovariance = std::decay_t<decltype(std::get<1>(beluga::estimate(states, weights)))>;
+
+  static_assert(std::is_same_v<State, EstimateState>);
+
+  struct Particle {
+    State state;
+    Weight weight;
+    Cluster cluster;
+
+    /// Convenient factory method to pass to `zip_with`.
+    static constexpr auto construct(const State& s, Weight w, Cluster c) { return Particle{s, w, c}; }
+  };
+
+  struct Estimate {
+    Weight weight;
+    EstimateState mean;
+    EstimateCovariance covariance;
+
+    /// Constructor used with `emplace_back`.
+    constexpr Estimate(Weight w, EstimateState m, EstimateCovariance c)
+        : weight{w}, mean{std::move(m)}, covariance{std::move(c)} {}
+
+    constexpr bool operator<(const Estimate& other) const { return weight < other.weight; }
+    constexpr bool operator<=(const Estimate& other) const { return weight <= other.weight; }
+    constexpr bool operator>(const Estimate& other) const { return weight > other.weight; }
+    constexpr bool operator>=(const Estimate& other) const { return weight >= other.weight; }
+    constexpr bool operator==(const Estimate& other) const { return weight == other.weight; }
+    constexpr bool operator!=(const Estimate& other) const { return weight != other.weight; }
+  };
+
+  // For each cluster found, estimate the mean and covariance of the states that belong to it.
+  const auto unique_clusters = clusters | ranges::to<std::unordered_set>;
+  auto estimates = std::vector<Estimate>{};
+  estimates.reserve(unique_clusters.size());
+
+  for (const auto cluster : unique_clusters) {
+    auto particles = ranges::views::zip_with(&Particle::construct, states, weights, clusters) |          //
```

Comment:
Yeah, this is not great. Just out of curiosity, do you think that sorting the indices and rearranging would be faster than simply sorting the container of particles after copying?

`chunk_by` was added in [`range-v3 0.12.0`](https://github.com/ericniebler/range-v3/releases/tag/0.12.0) and doesn't exist in `range-v3 0.10.0` which is distributed with Focal. There was a `group_by` view before but it had different semantics.

For reference: https://togithub.com/ericniebler/range-v3/issues/1637#issuecomment-901514674

Reply:
Attempted to address this here: https://github.com/Ekumen-OS/beluga/pull/275/commits/054dd023f753ccf2cd1f770c2b402cbe297c0116

Reply:
> Just out of curiosity, do you think that sorting the indices and rearranging would be faster than simply sorting the container of particles after copying?

It depends on how the sorting algorithm works, but probably the same. It would get better if we could take a set of indices from a range without copying, but that's a stretch.

> chunk_by was added in [range-v3 0.12.0](https://github.com/ericniebler/range-v3/releases/tag/0.12.0) only available in [Noble](https://packages.ubuntu.com/source/noble/range-v3). There was a group_by view before but it had different semantics.

Oh, bummer :(

I guess we could achieve the same with an `std::unordered_map<Cluster, std::vector<std::pair<State, Weight>>>` or similar.